### PR TITLE
Checking in the first cut of provider infrastructure, plus code refactoring

### PR DESCRIPTION
**Pull request by cybertyche**
This checkin has three primary goals:
- Establish the Provider DLL with some initial functionality
- Fix an issue with the Stitch operator where the payloads were not be compared correctly
- A substantial amount of code refactoring enabled by new VS 2019 features

### CODE REVIEW

File: Sources/Core/Microsoft.StreamProcessing/Microsoft.StreamProcessing.csproj [hunk 1]
```diff
@@ -1320,5 +1320,9 @@
       <DependentUpon>FlushPolicy.tt</DependentUpon>
     </None>
   </ItemGroup>
+
+<!--  <ItemGroup>
+    <ProjectReference Include="..\Microsoft.StreamProcessing.Provider\Microsoft.StreamProcessing.Provider.csproj" />
+  </ItemGroup> -->
```
Comment (peterfreiling) [inline Sources/Core/Microsoft.StreamProcessing/Microsoft.StreamProcessing.csproj]: Delete? Is this a placeholder? I would have expected the provider project to take a dependency on the main one, not the other way around?

File: Sources/Core/Microsoft.StreamProcessing/Operators/Shufflecast/ShufflecastPipe.cs [hunk 2]
```diff
@@ -165,7 +165,7 @@ public unsafe void OnNext(StreamMessage<TInnerKey, TSource> batch)
             }
         }
 
-        private void AddPunctuationOrLowWatermarkToAllBatches(StreamMessage<TInnerKey, TSource>[] batches, long syncTime, long otherTime, TInnerKey innerKey, int hash)
```
Comment (peterfreiling) [inline Sources/Core/Microsoft.StreamProcessing/Operators/Shufflecast/ShufflecastPipe.cs]: >long otherTime [](start = 122, length = 14)

This looks like a bug - we should be setting vother = otherTime

File: Sources/Core/Microsoft.StreamProcessing/Operators/Union/PartitionedUnionPipe.cs [hunk 3]
```diff
@@ -262,7 +258,7 @@ private void ProcessPendingEntries()
                         {
                             // Output the right tuple
                             OutputCurrentTuple(rightEntry);
-                            rightWorking.TryDequeue(out rightEntry);
+                            _ = rightWorking.TryDequeue(out _);
```
Comment (peterfreiling) [inline Sources/Core/Microsoft.StreamProcessing/Operators/Union/PartitionedUnionPipe.cs]: >_ =  [](start = 28, length = 4)

why assign to a discard?

File: Sources/Core/Microsoft.StreamProcessing/Microsoft.StreamProcessing.csproj (hunk 1 above)
Comment (cybertyche) [reply Sources/Core/Microsoft.StreamProcessing/Microsoft.StreamProcessing.csproj]: It's indeed a placeholder, and sadly, the direction is correct.

File: Sources/Core/Microsoft.StreamProcessing.Provider/IGroupedWindow.cs [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+// Licensed under the MIT License
+// *********************************************************************
+
+namespace Microsoft.StreamProcessing.Provider
+{
+    /// <summary>
+    ///
+    /// </summary>
+    /// <typeparam name="TKey"></typeparam>
+    /// <typeparam name="TPayload"></typeparam>
```
Comment (peterfreiling) [inline Sources/Core/Microsoft.StreamProcessing.Provider/IGroupedWindow.cs]: Can we fill these out? Probably best to not add more empty documentation if avoidable :)

File: Sources/Core/Microsoft.StreamProcessing/Operators/Shufflecast/ShufflecastPipe.cs (hunk 2 above)
Comment (cybertyche) [reply Sources/Core/Microsoft.StreamProcessing/Operators/Shufflecast/ShufflecastPipe.cs]: Understood - the real bug was that the parameter was not being used.

File: Sources/Core/Microsoft.StreamProcessing/Operators/Union/PartitionedUnionPipe.cs (hunk 3 above)
Comment (cybertyche) [reply Sources/Core/Microsoft.StreamProcessing/Operators/Union/PartitionedUnionPipe.cs]: Fixed, checking in a change momentarily.

File: Sources/Core/Microsoft.StreamProcessing/Operators/Stitch/StitchPipe.cs [hunk 5]
```diff
@@ -218,6 +218,7 @@ public override unsafe void OnNext(StreamMessage<TKey, TPayload> input)
                             this.batch[this.outputCount] = default;
                             this.batch.key.col[this.outputCount] = default;
                             this.batch.hash.col[this.outputCount] = 0;
+                            this.batch.bitvector.col[this.outputCount >> 6] |= 1L << (this.outputCount & 0x3f);
```
Comment (peterfreiling) [inline Sources/Core/Microsoft.StreamProcessing/Operators/Stitch/StitchPipe.cs]: >this.batch.bitvector.col[this.outputCount >> 6] |= 1L << (this.outputCount & 0x3f); [](start = 28, length = 83)

Is this the fix to Stitch you mention in the description?

Comment (cybertyche) [reply Sources/Core/Microsoft.StreamProcessing/Operators/Stitch/StitchPipe.cs]: Yes - payloads were being considered even if it was a punctuation.

**Comment by NickDarvey** [comment]
Is there an issue related to this work or somewhere I can read about it? (This is like IQbservable or IQueryable right?)

**Comment by cybertyche** [comment]
@NickDarvey Correct - this checkin is just a prototype so far, but I will file an issue on it next week. But it will be the IStreamable version of IQueryable, yes.

**Comment by cybertyche** [comment]
Come to think of it, I'll make sure it's up tomorrow. I'd love to have as many eyes (or hands) on it as possible.

**Comment by cybertyche** [comment]
@NickDarvey Issue #70 filed. Comments and contribution totally welcome.

**Pull request merged**